tests/various/gen_if_null.ys

```diff
@@ -0,0 +1,3 @@
+read_verilog gen_if_null.v
+design -save read
+prep
```

Comment:
These two lines aren't necessary for the test.
`design -save read` is copied from another test -- its purpose is to save a snapshot of the design under the `read` label so that we don't have to read it from disk again when we re-use it for a different test.
`prep` is part of synthesis, and isn't necessary for a test that just touches the Verilog frontend.

Amusingly, both iverilog and verilator give a warning:
```
$ iverilog -V
Icarus Verilog version 10.3 (stable) ()
<<SNIP>>
$ iverilog -s top gen_if_null.v 
gen_if_null.v:7: syntax error
gen_if_null.v:7: error: invalid module item.
gen_if_null.v:9: syntax error
gen_if_null.v:9: error: invalid module item.
```
```
$ verilator -V
Verilator 4.016 2019-06-16 rev UNKNOWN_REV
<<SNIP>>
$ verilator --lint-only --top-module top gen_if_null.v 
%Error: gen_if_null.v:5: syntax error, unexpected input
%Error: gen_if_null.v:6: syntax error, unexpected input
%Error: gen_if_null.v:11: syntax error, unexpected input
%Error: Exiting due to 3 error(s)
```

Reply:
I removed the `design -save read`. Thanks for explaining!

I think `prep`, or something like it is needed. In a prior implementation, I had made the frontend parse successfully, but the branches would be mis-associated. This test ensures the correct branch is taken by declaring the ports in the desired branches. It seems that yosys does not check the ports until I run `prep`. Perhaps there is a better choice for triggering this behavior?

Reply:
Can you recall what you had in the prior implementation?

`read_verilog` is responsible for generating the AST and for elaborating the design into a netlist (RTLIL), which requires port knowledge. Without this PR, it fails on `read_verilog` with:
```
 ./yosys -q gen_if_null.v 
gen_if_null.v:4: ERROR: syntax error, unexpected ';'
```

Reply:
I believe the overall patch for the naive implementation is simply:
```patch
diff --git a/frontends/verilog/verilog_parser.y b/frontends/verilog/verilog_parser.y
index 4a5aba79..8faede80 100644
--- a/frontends/verilog/verilog_parser.y
+++ b/frontends/verilog/verilog_parser.y
@@ -2533,7 +2533,7 @@ gen_stmt:
                ast_stack.back()->children.push_back(node);
                ast_stack.push_back(node);
                ast_stack.back()->children.push_back($3);
-       } gen_stmt_block opt_gen_else {
+       } gen_stmt_or_null opt_gen_else {
                SET_AST_NODE_LOC(ast_stack.back(), @1, @7);
                ast_stack.pop_back();
        } |
```

When built with this patch, `read_verilog` on the test file runs without issue, but running `prep` yields:
```
2.1. Executing HIERARCHY pass (managing design hierarchy).
ERROR: Module `test' referenced in module `top' in cell `test' has only 2 ports, requested port 3.
```

Reply:
Ooh, fascinating! Thanks for digging into it @zachjs -- it was more for my curiosity than it was doubting what you were doing.

So with your naive patch, it seems that `test` is elaborated into:
```
module test(y, z);
  input y;
  input z;
endmodule
```
which is a mismatch when it comes to `module test(x, y, z);` from the original input. I would have expected this to give an error, meaning you wouldn't need to resolve cell instantiations (done by the `hierarchy` pass) in order to see an error.

Reply:
Rather than relying on an instantiation, you can also consider checking for the existence of ports explicitly:
```
select -assert-count 1 i:x
select -assert-count 1 i:y
select -assert-count 1 i:z
```

Reply:
Thanks @eddiehung! That test procedure makes a lot more sense.